avro/schema.go

```diff
@@ -0,0 +1,113 @@
+package avro
+
+import (
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	schemaregistry "github.com/Landoop/schema-registry"
+	"github.com/linkedin/goavro"
+)
+
+type cachedCodec struct {
+	done  chan struct{}
+	codec *goavro.Codec
+	err   error
+}
+
+// SchemaCache connects to the Confluent schema registry and maintains
+// a cached versions of Avro schemas and codecs.
+type SchemaCache struct {
+	client *schemaregistry.Client
+
+	mu               sync.RWMutex
+	codecsBySchemaID map[int]*cachedCodec
+}
+
+// NewSchemaCache returns a new Cache instance
+func NewSchemaCache(url string) (*SchemaCache, error) {
+	client, err := schemaregistry.NewClient(url)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &SchemaCache{
+		codecsBySchemaID: make(map[int]*cachedCodec),
+		client:           client,
+	}
+	return c, nil
+}
+
+// getCodecForSchemaID returns a goavro codec for transforming data.
+func (c *SchemaCache) getCodecForSchemaID(schemaID int) (codec *goavro.Codec, err error) {
+	c.mu.RLock()
+	cc, ok := c.codecsBySchemaID[schemaID]
+	c.mu.RUnlock()
+	if ok {
+		<-cc.done
+		return cc.codec, cc.err
+	}
+
+	// Codec is not cached, grab exclusive lock and ensure no other
+	// goroutine started the process in-between.
+	c.mu.Lock()
+	cc, ok = c.codecsBySchemaID[schemaID]
+	if ok {
+		// Another goroutine began fetching schema and codec.
+		c.mu.Unlock()
+		<-cc.done
+		return cc.codec, cc.err
+	}
+
+	// Create the cachedCodec with a promise of a future value.
+	cc = &cachedCodec{done: make(chan struct{})}
+	c.codecsBySchemaID[schemaID] = cc
+	c.mu.Unlock()
+
+	defer func() {
```

Comment:
to be honest i have trouble understanding the purpose of this construct, also with the channel and the error. why don't we just directly set the codec here instead of doing this in defer. can you elaborate?
also i think a normal mutex for this function is sufficient. i doubt we'll have performance issues.

Reply:
Apologies, this construct is my handiwork so I'll try to answer it instead of @tzetter.

The purpose of the defer is to assign the returned values to `cc` (`cachedCodec`) no matter which of the three code paths are taken by accessing the variables from the return signature (`(codec *goavro.Codec, err error)`). If we did not we would need to `cc.err = err; close(cc.done);` in both of the `if err != nil` check below. And in case of no error we would do `cc.codec = codec; close(cc.done);`. So the construct just ensures that we always finalize the `cachedCodec`.

I realize that is not easy to read in a sentence, so I'll illustrate how this construct would look without the defer:

```go
	schema, err := c.client.GetSchemaById(schemaID)
	if err != nil {
		cc.err = err
		close(cc.done)
		return nil, err
	}

 	codec, err = goavro.NewCodec(schema)
	if err != nil {
		cc.err = err
		close(cc.done)
		return nil, err
	}

	cc.codec = codec
	close(cc.done)
 	return codec, nil
```

The purpose of the done channel is to inform other goroutines that `cc.codec` and `cc.err` are ready to be read. Let's say one routine starts fetching the codec and has inserted its `&cachedCodec{done: make(chan struct{})}`, any other routine that tries to the same codec will end up waiting on `<-cc.done`. Thus, when it closes all routines can carry on and the values will be present. Kind of like a JavaScript promise, or similar construct.

> also i think a normal mutex for this function is sufficient. i doubt we'll have performance issues.

As in `sync.Mutex` and holding it for the entire duration? I suppose that would work as well, although it would force the other partition consumers to wait whenever there is an uncached codec. It's probably acceptable though.

Reply:
i guess it's a reasonable solution, even if the readability is not ideal. thanks for the explanation